### PR TITLE
Closed classes 2

### PR DESCRIPTION
**Pull request by RikkiGibson**
Test plan: #81039 

- Add `NamedTypeSymbol.IsClosed` API
- Synthesize `ClosedAttribute` on closed types
- Synthesize `CompilerFeatureRequiredAttribute` on constructors of closed types
- Add same-assembly subtype restriction
- Add subtype type parameter restriction

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs [hunk 1]
```diff
@@ -1520,7 +1521,7 @@ private DiagnosticInfo DeriveCompilerFeatureRequiredDiagnostic()
             var diag = PEUtilities.DeriveCompilerFeatureRequiredAttributeDiagnostic(
                 this, containingModule, Handle,
                 allowedFeatures: MethodKind == MethodKind.Constructor ?
-                    CompilerFeatureRequiredFeatures.RequiredMembers :
+                    CompilerFeatureRequiredFeatures.RequiredMembers | CompilerFeatureRequiredFeatures.ClosedClasses :
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs]: >CompilerFeatureRequiredFeatures.RequiredMembers | CompilerFeatureRequiredFeatures.ClosedClasses [](http://example.com/codeflow?start=20&length=95)

Are we allowing multiple applications of CompilerFeatureRequired attribute? For example, is it possible that we might need to mark a constructor with both feature? If so, how would we do this?  #Closed

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs]: >CompilerFeatureRequiredFeatures.ClosedClasses [](http://example.com/codeflow?start=70&length=45)

Is presence of this attribute going affect consumption of the constructor in a non-closed class? If not, then perhaps we should not recognize the attribute as valid on a constructor like that. #Closed

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 2]
```diff
@@ -877,14 +877,16 @@ internal override ManagedKind GetManagedKind(ref CompoundUseSiteInfo<AssemblySym
 
         public override bool IsSealed => HasFlag(DeclarationModifiers.Sealed);
 
-        public override bool IsAbstract => HasFlag(DeclarationModifiers.Abstract);
+        public override bool IsAbstract => HasFlag(DeclarationModifiers.Abstract) || HasFlag(DeclarationModifiers.Closed);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: >HasFlag(DeclarationModifiers.Abstract) || HasFlag(DeclarationModifiers.Closed); [](http://example.com/codeflow?start=43&length=79)

This feels wrong. When we are processing modifiers we should set `DeclarationModifiers.Abstract` flag if there is a `closed` modifier. The flags are not meant to reflect what was in syntax, they are meant to reflect result of binding the syntax. #Closed

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs]: The definition of `CompilerFeatureRequiredAttribute` permits multiple applications. I added `RequiredMembers_` tests for these cases. I think testing the consumption of a "closed class with required members" from metadata is missing, I will add that.

I can also ensure we have tests which handle multiple applications of `CompilerFeatureRequiredAttribute` being present, with e.g mix of known and unknown features, and ensure the expected errors are reported when using the symbols. My expectation was that if *any* of the attributes represents an unknown required feature, then, an error would be reported. I could be wrong about that, so, if that ends up being the case will revisit how to encode this.

The use of CompilerFeatureRequiredAttribute in this way, may also require LDM approval, I'm going to expand on https://github.com/dotnet/csharplang/blob/main/proposals/closed-hierarchies.md#open-questions and get an agenda slot for it.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs]: > Is presence of this attribute going affect consumption of the constructor in a non-closed class?

The goal is to ensure that when an old compiler encounters a closed class in metadata, they are able to use the class and its members in general, except they can't declare a derived type from it. I thought that the constructor of an abstract class could effectively only be used in a constructor initializer, so, this approach would be effective toward restricting only the unwanted use.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: For example, interface instance members without modifiers have `DeclarationModifiers.Abstract` flag set. Because they are abstract by default

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs]: I think we should understand the feature only in the setting in which we understand it. For example, we do not allow `RequiredMembers` on all methods. Only on constructors, where we understand what the requirement means

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs]: Basically this version of compiler knows what this requirement means on a constructor of a closed class, but not on any other constructor

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs]: Or compiler should block usage of the constructor regardless whether it is a closed class constructor

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEMethodSymbol.cs]: I guess I would be fine with either approach where the feature requirement is not silently ignored

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs [hunk 3]
```diff
@@ -404,6 +404,46 @@ static bool containsOnlyOblivious(TypeSymbol type)
...
                 }
+
+                if (baseType.IsClosed)
+                {
+                    if ((object)baseType.ContainingModule != DeclaringCompilation.SourceModule)
+                    {
+                        // https://github.com/dotnet/csharplang/blob/3e15888c804dc632479c94589dcd02c341fd0a4f/proposals/closed-hierarchies.md#same-assembly-restriction
+                        // > If a class in one assembly is declared closed then it is an error to directly derive from it in another assembly.
+                        // Note that while the spec language is in terms of *other assembly*, the same justification applies for cross-module scenario.
+                        // i.e. the pattern matching in the declaring module needs to know the full set of subtypes, so a referencing module must not introduce new subtypes.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:415]: >// i.e. the pattern matching in the declaring module needs to know the full set of subtypes, so a referencing module must not introduce new subtypes. [](http://example.com/codeflow?start=24&length=149)

Strongly support this line of thinking. #Closed

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs [hunk 4]
```diff
@@ -230,6 +230,13 @@ public CallerArgumentExpressionAttribute(string parameterName)
             }
             """;
 
+        protected static readonly string ClosedAttributeDefinition = """
+            namespace System.Runtime.CompilerServices
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:234]: >namespace System.Runtime.CompilerServices [](http://example.com/codeflow?start=12&length=41)

Consider adding this attribute declaration to the speclet. And, since we are planning to add it to BCL, once we are confident in the shape, we should submit an API proposal for BCL and pass an API review.  #Closed

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs [hunk 5]
```diff
@@ -30,47 +31,148 @@ public void LangVersion()
...
             """;
-        var source2 = """
-            public class D : C { }
-            """;
 
-        var comp = CreateCompilation([source1, source2]);
-        comp.VerifyEmitDiagnostics();
+        var comp1 = CreateCompilation(source1, targetFramework: TargetFramework.Net100);
+        comp1.VerifyEmitDiagnostics(
+            // (1,21): error CS0656: Missing compiler required member 'System.Runtime.CompilerServices.ClosedAttribute..ctor'
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:58]: >// (1,21): error CS0656: Missing compiler required member 'System.Runtime.CompilerServices.ClosedAttribute..ctor' [](http://example.com/codeflow?start=12&length=113)

It looks like scenario when `CompilerFeatureRequiredAttribute` is missing is not covered. #Closed

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs [hunk 6]
```diff
@@ -113,25 +219,591 @@ public void Abstract_01()
...
+            """;
+        var vbComp = CreateVisualBasicCompilation("Program", source2, referencedCompilations: [comp1], referencedAssemblies: comp1.References);
+
+        // note: the multi-line strings in the diagnostic arguments not behaving well when pasting in the VerifyDiagnostics baseline.
+        // Just verifying the things we are interested in directly instead.
+        var diagnostics = vbComp.GetDiagnostics();
+        Assert.Equal(2, diagnostics.Length);
+        Assert.Equal(36954 /*ERRID.ERR_BadOverloadCandidates2*/, diagnostics[0].Code);
+        Assert.Equal(2, diagnostics[0].Arguments.Count);
+        AssertEx.AssertEqualToleratingWhitespaceDifferences("New", diagnostics[0].Arguments[0]!.ToString());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs]: >! [](http://example.com/codeflow?start=94&length=1)

Consider suppressing nullable analysis at the top. Test code doesn't need to handle all possible scenarios. If you run test ones and it doesn't crash, nullability of values is not expected to change during future runs.  #Closed

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs [hunk 7]
```diff
@@ -113,25 +219,591 @@ public void Abstract_01()
...
+        //error BC0000:
+        //        'Public Overloads Sub New()': 'Public Overloads Sub New()' requires compiler feature 'ClosedClasses', which is not supported by this version of the Visual Basic compiler.
+        //        'Public Overloads Sub New(i As Integer)': 'Public Overloads Sub New(i As Integer)' requires compiler feature 'ClosedClasses', which is not supported by this version of the Visual Basic compiler.
+        Assert.Contains(expectedSubstring: "ClosedClasses", actualString: diagnostics[0].Arguments[1]!.ToString());
+        Assert.Equal(": (4,15)-(4,18)", diagnostics[0].Location.GetLineSpan().ToString());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs]: >: (4,15)-(4,18) [](http://example.com/codeflow?start=22&length=15)

It is hard to tell what line it is. Consider using line directive with distinct enough value #Closed

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs [hunk 8]
```diff
@@ -113,25 +219,591 @@ public void Abstract_01()
...
+
+        Assert.Equal(36954 /*ERRID.ERR_BadOverloadCandidates2*/, diagnostics[1].Code);
+        Assert.Equal(2, diagnostics[1].Arguments.Count);
+        AssertEx.AssertEqualToleratingWhitespaceDifferences("New", diagnostics[1].Arguments[0]!.ToString());
+        //error BC0000:
+        //        'Public Overloads Sub New()': 'Public Overloads Sub New()' requires compiler feature 'ClosedClasses', which is not supported by this version of the Visual Basic compiler.
+        //        'Public Overloads Sub New(i As Integer)': 'Public Overloads Sub New(i As Integer)' requires compiler feature 'ClosedClasses', which is not supported by this version of the Visual Basic compiler.
+        Assert.Contains(expectedSubstring: "ClosedClasses", actualString: diagnostics[1].Arguments[1]!.ToString());
+        Assert.Equal(": (8,15)-(8,18)", diagnostics[1].Location.GetLineSpan().ToString());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs]: >(8,15)-(8,18) [](http://example.com/codeflow?start=24&length=13)

Here as well #Closed

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 9]
```diff
@@ -1172,7 +1172,8 @@ protected sealed override void DecodeWellKnownAttributeImpl(ref DecodeWellKnownA
                 | ReservedAttributes.NativeIntegerAttribute
                 | ReservedAttributes.CaseSensitiveExtensionAttribute
                 | ReservedAttributes.RequiredMemberAttribute
-                | ReservedAttributes.ExtensionMarkerAttribute))
+                | ReservedAttributes.ExtensionMarkerAttribute
+                | ReservedAttributes.ClosedAttribute))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: >ReservedAttributes.ClosedAttribute [](http://example.com/codeflow?start=18&length=34)

I think we should disallow the attribute in all other contexts as well. #Closed

**Comment by AlekseyTs** [comment]
Done with review pass (commit 6)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 9 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: I added AttributeTargets to the declaration, this is also part of what we will propose to BCL. Does this address the concern, or is there also a need to e.g. ensure we block use of a ClosedAttribute which doesn't have this particular AttributeTargets?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: Yes, I meant to add this suppression for all other attribute contexts, regardless what targets are valid for the attribute

**Comment by AlekseyTs** [comment]
Done with review pass (commit 11)

File: src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousManager.TypeOrDelegatePublicSymbol.cs [hunk 10]
```diff
@@ -138,6 +138,8 @@ internal sealed override bool HasSpecialName
 
             internal override bool HasDeclaredRequiredMembers => false;
 
+            internal override bool IsClosed => false;
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/AnonymousTypes/PublicSymbols/AnonymousManager.TypeOrDelegatePublicSymbol.cs]:
```suggestion
            internal sealed override bool IsClosed => false;
```

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 11]
```diff
@@ -389,6 +389,11 @@ private DeclarationModifiers MakeModifiers(TypeKind typeKind, BindingDiagnosticB
                 case TypeKind.Interface:
                     mods |= DeclarationModifiers.Abstract;
                     break;
+                case TypeKind.Class:
+                    if ((mods & DeclarationModifiers.Closed) != 0)
+                        mods |= DeclarationModifiers.Abstract;
+
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:395]: nit: remove blank line?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 12]
```diff
@@ -1892,6 +1899,13 @@ protected void AfterMembersChecks(BindingDiagnosticBag diagnostics)
                 compilation.EnsureIsReadOnlyAttributeExists(diagnostics, location, modifyCompilation: true);
             }
 
+            if (this.IsClosed)
+            {
+                // Ensure necessary attributes are present
+                _ = Binder.GetWellKnownTypeMember(DeclaringCompilation, WellKnownMember.System_Runtime_CompilerServices_ClosedAttribute__ctor, diagnostics, GetFirstLocation());
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1905]: Are we not going to synthesize our own ClosedAttribute definition if it doesn't exist?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs (hunk 9 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: > disallow the attribute in all other contexts as well.

We aren't doing that for other attributes though. It seems we disallow reserved attributes only on the declaration kinds where we recognize them. Perhaps so that if we start supporting that attribute on more declarations, this would allow people that use older compilers to specify them manually?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: We weren't doing this in a consistent manner, I guess. And I feel like that was due to an oversight rather than by intent. For example `IsByRefLikeAttribute` is suppressed not every where but in most targets (a couple were missed). And there are probably other examples of suppressing in contexts that the attribute is not meant to be applied. I think "reserved" attributes should be suppressed everywhere. And we probably should change the implementation code to require a code change to allow a reserved attributes at a specific context, rather than require a change to forbit is at a specific context. We can achieve this by always passing the value with all bits set (-1), and, for backward compatibility, clear the bits that are currently allowed. This way, going forward any new reserved attribute will be automatically forbidden everywhere. No additional work would be needed.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs]: Or even simpler, change the meaning of the parameter to represent attributes that are allowed, and fix-up the call sites accordingly.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 12 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1905]: That's right, we won't synthesize a definition. User needs to provide it. This is in keeping with `RequiredMemberAttribute` and `CompilerFeatureRequiredAttribute` behavior.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 11 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:395]: My personal preference is to keep a blank line after an if statement which lacks a block body

**Comment by AlekseyTs** [comment]
@RikkiGibson It looks like there are legitimate test failures

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs [hunk 13]
```diff
@@ -2368,7 +2368,26 @@ private void DecodeWellKnownAttribute(ref DecodeWellKnownAttributeArguments<Attr
             int signature;
             var diagnostics = (BindingDiagnosticBag)arguments.Diagnostics;
 
-            if (attribute.IsTargetAttribute(AttributeDescription.InternalsVisibleToAttribute))
+            if (attribute is SourceAttributeData
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs:2371]: In absence of the `attribute is SourceAttributeData` check, I was seeing an assertion failure in test [CommandLineTests.AddModuleWithExtensionMethod](https://github.com/dotnet/roslyn/blob/7abf171ce8515ec5481a4aa8e654f421198b9a2c/src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs#L5944).

It looks like, effectively, a PE attribute can be included in the attributes for a source assembly, via a netmodule, and this was failing an assertion in `ReportExplicitUseOfReservedAttributes`.

I think in this case, we should assume that for a PE attribute, the use of reserved attributes was already checked when compiling the module whose PE attribute we are including. So, prepending the `is SourceAttributeData` check, feels like a reasonable solution to me.

Would be interested to know any others thoughts or insight about this.

<details>
    <summary>Stack trace</summary>

```
System.InvalidOperationException : attribute is SourceAttributeData
at Microsoft.CodeAnalysis.ThrowingTraceListener.Fail(String message, String detailMessage) (src/Compilers/Test/Core/ThrowingTraceListener.cs#L26,0)
at System.Diagnostics.TraceInternal.Fail(String message, String detailMessage)
at System.Diagnostics.Debug.Fail(String message, String detailMessage)
at Microsoft.CodeAnalysis.CSharp.Symbol.ReportExplicitUseOfReservedAttributes(DecodeWellKnownAttributeArguments`3& arguments, ReservedAttributes permitted) (src/Compilers/CSharp/Portable/Symbols/Symbol.cs#L1543,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.DecodeWellKnownAttribute(DecodeWellKnownAttributeArguments`3& arguments, Int32 index, Boolean isFromNetModule) (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L2371,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.ValidateAttributeUsageAndDecodeWellKnownAttributes(ImmutableArray`1 attributesFromNetModules, ImmutableArray`1 netModuleNames, BindingDiagnosticBag diagnostics) (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L1352,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.LoadAndValidateNetModuleAttributes(CustomAttributesBag`1& lazyNetModuleAttributesBag) (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L1380,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.GetNetModuleAttributesBag() (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L1539,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.GetNetModuleDecodedWellKnownAttributeData() (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L1546,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.get_AssemblyVersionAttributeSetting() (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L400,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.ComputeIdentity() (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L2181,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.get_Identity() (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L199,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.ValidateIVTPublicKeys(BindingDiagnosticBag diagnostics) (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L733,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.ValidateAttributeSemantics(BindingDiagnosticBag diagnostics) (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L591,0)
at Microsoft.CodeAnalysis.CSharp.Symbols.SourceAssemblySymbol.ForceComplete(SourceLocation locationOpt, Predicate`1 filter, CancellationToken cancellationToken) (src/Compilers/CSharp/Portable/Symbols/Source/SourceAssemblySymbol.cs#L922,0)
at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSourceDeclarationDiagnostics(SyntaxTree syntaxTree, Nullable`1 filterSpanWithinTree, Func`4 locationFilterOpt, Predicate`1 symbolFilter, CancellationToken cancellationToken) (src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L3315,0)
at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnosticsWithoutSeverityFiltering(CompilationStage stage, Boolean includeEarlierStages, BindingDiagnosticBag builder, Predicate`1 symbolFilter, CancellationToken cancellationToken) (src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L3089,0)
at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnostics(CompilationStage stage, Boolean includeEarlierStages, DiagnosticBag diagnostics, Predicate`1 symbolFilter, CancellationToken cancellationToken) (src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L3014,0)
at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnostics(CompilationStage stage, Boolean includeEarlierStages, DiagnosticBag diagnostics, CancellationToken cancellationToken) (src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs#L3007,0)
at Microsoft.CodeAnalysis.CommonCompiler.CompileAndEmit(TouchedFileLogger touchedFilesLogger, Compilation& compilation, ImmutableArray`1 analyzers, ImmutableArray`1 generators, ImmutableArray`1 additionalTextFiles, AnalyzerConfigSet analyzerConfigSet, ImmutableArray`1 sourceFileAnalyzerConfigOptions, ImmutableArray`1 embeddedTexts, DiagnosticBag diagnostics, ErrorLogger errorLogger, CancellationToken cancellationToken, CancellationTokenSource& analyzerCts, AnalyzerDriver& analyzerDriver, Nullable`1& generatorTimingInfo) (src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L1210,0)
at Microsoft.CodeAnalysis.CommonCompiler.RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, CancellationToken cancellationToken) (src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L949,0)
at Microsoft.CodeAnalysis.CommonCompiler.Run(TextWriter consoleOutput, CancellationToken cancellationToken) (src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L770,0)
at Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests.CommandLineTests.AddModuleWithExtensionMethod() (src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs#L5956,0)
```
</details>

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEmbeddedAttributeSymbol.cs [hunk 14]
```diff
@@ -70,6 +70,8 @@ public SynthesizedEmbeddedAttributeSymbolBase(
 
         internal override bool HasDeclaredRequiredMembers => false;
 
+        internal override bool IsClosed => false;
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEmbeddedAttributeSymbol.cs]:
```suggestion
        internal sealed override bool IsClosed => false;
```

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 15]
```diff
@@ -8247,4 +8247,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_ClosedSealedStatic" xml:space="preserve">
     <value>'{0}': a closed type cannot be sealed or static</value>
   </data>
+  <data name="ERR_ClosedBaseTypeBaseFromOtherAssembly" xml:space="preserve">
+    <value>'{0}': cannot use a closed type '{1}' from another assembly as a base type.</value>
+  </data>
+  <data name="ERR_UnderspecifiedClosedSubtype" xml:space="preserve">
+    <value>'{0}': The type parameter '{1}' must be referenced in the base type '{2}' because the base type is closed.</value>
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:8254]: The casing of the first letter after `'{0}': ` seems inconsistent - it's uppercase here but lowercase in the other two error messages above.

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs [hunk 16]
```diff
@@ -30,47 +33,173 @@ public void LangVersion()
             closed class C { }
             """;
 
-        var comp = CreateCompilation(source, parseOptions: TestOptions.Regular14);
+        var comp = CreateCompilation([source, ClosedAttributeDefinition], parseOptions: TestOptions.Regular14, targetFramework: TargetFramework.Net100);
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:36]: Why do we need target framework .NET 10?

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs [hunk 17]
```diff
@@ -113,25 +246,752 @@ public void Abstract_01()
...
+            } // end of class C
+            """;
+
+        var ilComp = CompileIL(il);
+        var source1 = """
+            public class D : C { }
+            """;
+        var comp1 = CreateCompilation(source1, references: [ilComp], targetFramework: TargetFramework.Net100);
+        comp1.VerifyEmitDiagnostics(
+            // (1,14): error CS9041: 'C.C()' requires compiler feature 'ClosedClasses', which is not supported by this version of the C# compiler.
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:555]: Is this error expected? It seems a bit misleading since ClosedClasses feature is supported by the current compiler.

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs [hunk 18]
```diff
@@ -113,25 +246,752 @@ public void Abstract_01()
...
+            .assembly extern System.Runtime { .ver 10:0:0:0 .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A) }
+
+            .class public auto ansi abstract beforefieldinit C
+                extends [System.Runtime]System.Object
+            {
+                // Methods
+                .method family hidebysig specialname rtspecialname
+                    instance void .ctor () cil managed
+                {
+                    .custom instance void [System.Runtime]System.Runtime.CompilerServices.CompilerFeatureRequiredAttribute::.ctor(string) = (
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:535]: Do we actually need ClosedAttribute if we are adding CompilerFeatureRequired("ClosedClasses") everywhere anyway? Perhaps we could save on metadata size if we used only the latter.

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs [hunk 19]
```diff
@@ -113,25 +246,752 @@ public void Abstract_01()
...
+                        01 00 0d 43 6c 6f 73 65 64 43 6c 61 73 73 65 73
+                        00 00
+                    )
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:538]: Btw, it looks like the use of CompilerFeatureRequired is not mentioned in the speclet. Should it be updated, or is it considered a compiler implementation detail (although I guess it could be mentioned anyway for clarity).

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:538]: The approach of using this attribute to block unwanted usage is going to LDM.

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs [hunk 20]
```diff
@@ -1531,86 +1531,92 @@ internal enum ReservedAttributes
             RefSafetyRulesAttribute = 1 << 13,
             RequiresLocationAttribute = 1 << 14,
             ExtensionMarkerAttribute = 1 << 15,
+            ClosedAttribute = 1 << 16,
         }
 
-        internal bool ReportExplicitUseOfReservedAttributes(in DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments, ReservedAttributes reserved)
+        // PROTOTYPE(cc): Remove unnecessary 'permitted' flags from call sites of this method
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs:1537]: >// PROTOTYPE(cc): Remove unnecessary 'permitted' flags from call sites of this method [](http://example.com/codeflow?start=8&length=85)

Could you provide more details on this pending work item? #Closed

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Symbol.cs:1537]: I made a mechanical change at the call sites to provide 'permitted' flags which are equivalent to the previous 'reserved' flags. I suspect that many of the permitted flags, really shouldn't be passed, and the usage of certain attributes should be blocked instead on various symbol kinds. I wanted to defer working out which flags to delete at call sites, which test coverage to add, etc. until after this PR or possibly until after feature merge.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Symbol.cs:1537]: Makes sense to not tie this to the feature work.

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs (hunk 18 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:535]: I'll bring this to LDM.

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs (hunk 16 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:36]: I preferred to test it this way as it represents the supported configuration, reduces number of polyfilled attribute definitions I need to pass, etc.

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs (hunk 17 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:555]: Yes, see thread https://github.com/dotnet/roslyn/pull/82052#discussion_r2742781080

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 15 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:8254]: I would prefer to address this in a future PR

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs (hunk 17 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:555]: I read that thread to mean that we expect _some_ error in this case, but still this particular error seems a bit misleading to me. Although it's such an edge case scenario that it's probably fine.

File: src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs (hunk 16 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/CSharp15/ClosedClassesTests.cs:36]: Consider adding at least one test with the default target framework (netstandard2.0, I believe) to demonstrate the feature works there too.
